### PR TITLE
Unordered PARTITION BY column implementation (to prevent pipeline breaking)

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#6035](https://github.com/apache/arrow-datafusion/issues/6035).

# Rationale for this change

Current `BoundedWindowAggExec` implementation assumes that its input is ordered both by PARTITION and ORDER BY columns. This enables us to calculate partition separation points easily and fast. However, this assumption results in some queries to break pipeline. Even if there is no theoretical reason to break pipeline. 

# What changes are included in this PR?

This PR extends functionality to support above mentioned cases. Consider query below
```sql
SELECT SUM(inc_col) OVER(PARTITION BY unsorted_col ORDER BY low_card_col1 ASC ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING) as sum1
FROM annotated_data2
```
where `low_card_col1` is ascending sorted. There is no fundamental reason for above query to sort table according `unsorted_col` column (This helps with partition calculation). When input source is marked as `infinite`, this PR would turn physical plan for above query from
```sql
"ProjectionExec: expr=[SUM(annotated_data2.inc_col) PARTITION BY [annotated_data2.unsorted_col] ORDER BY [annotated_data2.low_card_col1 ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING@4 as sum1]",
"  BoundedWindowAggExec: wdw=[SUM(annotated_data2.inc_col): Ok(Field { name: \"SUM(annotated_data2.inc_col)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(2)), end_bound: Following(UInt64(1)) }], mode=[Sorted]",
"    SortExec: expr=[unsorted_col@3 ASC NULLS LAST,low_card_col1@0 ASC NULLS LAST]",
"      CsvExec: files={1 group: [[table_path]]}, has_header=true, limit=None, projection=[low_card_col1, low_card_col2, inc_col, unsorted_col]",

```
to
```sql
"ProjectionExec: expr=[SUM(annotated_data2.inc_col) PARTITION BY [annotated_data2.unsorted_col] ORDER BY [annotated_data2.low_card_col1 ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING@4 as sum1]",
"  BoundedWindowAggExec: wdw=[SUM(annotated_data2.inc_col): Ok(Field { name: \"SUM(annotated_data2.inc_col)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(2)), end_bound: Following(UInt64(1)) }], mode=[Linear]",
"    CsvExec: files={1 group: [[table_path]]}, has_header=true, limit=None, projection=[low_card_col1, low_card_col2, inc_col, unsorted_col]",

```

# Are these changes tested?
Existing fuzzy tests are extended to test this case. Also new tests are added.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -21,17 +21,8 @@
 use ahash::RandomState;
 
 use arrow::{
-    array::{
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:24]: One of the util code under this file is moved to the `hash_utils.rs` All of the changes in this file and `hash_utils.rs` comes from code movement.

File: datafusion/core/tests/sqllogictests/test_files/window.slt [hunk 2]
```diff
@@ -361,7 +361,7 @@ SortPreservingMergeExec: [b@0 ASC NULLS LAST]
       AggregateExec: mode=FinalPartitioned, gby=[b@0 as b], aggr=[MAX(d.a), MAX(d.seq)]
         AggregateExec: mode=Partial, gby=[b@2 as b], aggr=[MAX(d.a), MAX(d.seq)]
           ProjectionExec: expr=[ROW_NUMBER() PARTITION BY [s.b] ORDER BY [s.a ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@2 as seq, a@0 as a, b@1 as b]
-            BoundedWindowAggExec: wdw=[ROW_NUMBER(): Ok(Field { name: "ROW_NUMBER()", data_type: UInt64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int64(NULL)), end_bound: CurrentRow }]
+            BoundedWindowAggExec: wdw=[ROW_NUMBER(): Ok(Field { name: "ROW_NUMBER()", data_type: UInt64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Int64(NULL)), end_bound: CurrentRow }], mode=[Sorted]
```
Comment (mustafasrepo) [inline datafusion/core/tests/sqllogictests/test_files/window.slt:364]: In the print for `BoundedWindowAggExec` I added PartitionSearch mode. The reason is that `BoundedWindowAggExec` can produce correct result in `Linear` mode even if it should have been working in the `Sortted` mode. If this behaviour changes we may not recognize it. With this change we can prevent it from happenning. Most of the changes in this file comes from this change or addition of new tests

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 3]
```diff
@@ -270,6 +311,485 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+        // partition index corresponds to index of the partition in the vector `partition_indices`
+        let mut row_map: RawTable<(u64, usize, usize)> = RawTable::with_capacity(0);
+        // `partition_indices` store vector of tuple. First entry is the partition key (unique for each partition)
+        // second entry is indices of the rows which partition is constructed in the input record batch
+        // Third entry is the how many outputs are generated for the partition.
+        let mut partition_indices: Vec<(Vec<ScalarValue>, Vec<usize>)> = vec![];
+        assert_eq!(input_buffer.num_rows(), self.input_buffer_hashes.len());
+        for (row_idx, hash) in self.input_buffer_hashes.iter().enumerate() {
+            let entry = row_map.get_mut(*hash, |(_hash, group_idx, _n_out)| {
+                let row = get_row_at_idx(&partition_by_columns, row_idx).unwrap();
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:575]: Since closure doesn't return error. I couldn't remove this `.unwrap`. If you are aware any method to remove it. I am happy to change it.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 4]
```diff
@@ -270,6 +311,485 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+    ) -> Result<Vec<(PartitionKey, Vec<usize>)>> {
+        let mut batch_hashes = vec![0; batch.num_rows()];
+        create_hashes(columns, &self.random_state, &mut batch_hashes)?;
+        self.input_buffer_hashes.extend_from_slice(&batch_hashes);
+        let mut row_map: RawTable<(u64, usize)> = RawTable::with_capacity(0);
+        // res stores indices for each partition.
+        let mut res: Vec<(PartitionKey, Vec<usize>)> = vec![];
+        for (row_idx, hash) in batch_hashes.into_iter().enumerate() {
+            let entry = row_map.get_mut(hash, |(_hash, group_idx)| {
+                equal_rows(row_idx, res[*group_idx].1[0], columns, columns, true).unwrap()
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:542]: Since closure doesn't return error. I couldn't remove this `.unwrap`. If you are aware any method to remove it. I am happy to change it.

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 5]
```diff
@@ -925,83 +1093,252 @@ mod tests {
...
-                },
-            };
-            let (is_aligned, reverse) =
-                check_alignment(&schema, &physical_ordering, &required_ordering);
-            assert_eq!(is_aligned, is_aligned_expected);
-            assert_eq!(reverse, reverse_expected);
+        let exec_unbounded = csv_exec_sorted(&test_schema, sort_exprs.clone(), true);
+        let exec_bounded = csv_exec_sorted(&test_schema, sort_exprs, false);
+
+        let test_cases = vec![
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:1110]: I have written edge cases in tests. However, if you want some test case to be tested also. Please feel free to add or try new tests to see if it works as your expectations

**Comment by mustafasrepo** [comment]
@ozankabak BTW, I have resolved merge conflict. It is ready for further review.

**Comment by ozankabak** [comment]
Let's de-conflict this and I will start reviewing

File: datafusion/core/src/physical_optimizer/pipeline_fixer.rs [hunk 6]
```diff
@@ -163,37 +159,62 @@ fn hash_join_convert_symmetric_subrule(
...
+                parent.clone(),
+                parent.schema(),
+                exec.partition_keys.clone(),
+                exec.partition_search_mode.clone(),
+                true,
+            );
+            return Some(new_exec.map(|exec| {
+                input.plan = Arc::new(exec) as _;
+                input
+            }));
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/pipeline_fixer.rs:212]: The only change you seem to making to `exec` here is to set the the `source_unbounded` attribute. You don't touch `partition_search_mode`, seems like you are leaving it to `EnforceSorting`.

Was this your intent? Is the only purpose of this rule to set the `source_unbounded` attribute? If so, the name of the rule is misleading.

BTW, I have some other comments about the `source_unbounded` attribute, you can find them in other places.

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 7]
```diff
@@ -573,38 +575,53 @@ fn analyze_window_sort_removal(
...
         }
         // We can not skip the sort, or window reversal requirements are not
         // uniform; then sort removal is not possible -- we immediately return.
         return Ok(None);
     }
-    let new_window_expr = if needs_reverse.unwrap() {
+
+    let (should_reverse, partition_search_mode) = search_flags;
+    if searching
+        || !(source_unbounded || partition_search_mode == PartitionSearchMode::Sorted)
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:618]: I don't fully understand the `!(source_unbounded || partition_search_mode == PartitionSearchMode::Sorted)` condition here -- please explain to me in our meeting.

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 8]
```diff
@@ -632,8 +649,21 @@ fn analyze_window_sort_removal(
                 new_child,
                 new_schema,
                 partition_keys.to_vec(),
+                partition_search_mode,
+                source_unbounded,
             )?) as _
         } else {
+            if partition_search_mode != PartitionSearchMode::Sorted {
+                // For `WindowAggExec` to work correctly PARTITION BY columns should be sorted.
+                // Hence if `PartitionSearchMode` is not `Sorted` we should satisfy required ordering.
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:658]: Should the comment read "... is ..." instead of "... is not ..."?

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 9]
```diff
@@ -762,16 +792,11 @@ fn get_sort_exprs(sort_any: &Arc<dyn ExecutionPlan>) -> Result<&[PhysicalSortExp
     }
 }
 
-/// Compares physical ordering and required ordering of all `PhysicalSortExpr`s
-/// to decide whether a `SortExec` before a `WindowAggExec` can be removed.
-/// A `None` return value indicates that we can remove the sort in question.
-/// A `Some(bool)` value indicates otherwise, and signals whether we need to
-/// reverse the ordering in order to remove the sort in question.
 fn can_skip_sort(
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:795]: Please write explanatory doc-comments (`///`) for this function.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 10]
```diff
@@ -26,39 +26,78 @@ use crate::physical_plan::expressions::PhysicalSortExpr;
...
+    /// None of the columns among the partition columns is ordered.
+    Linear,
+    /// Some columns of the partition columns are ordered but not all
+    PartiallySorted(Vec<usize>),
+    /// All Partition columns are ordered (Also empty case)
+    Sorted,
+}
+
+impl PartialOrd for PartitionSearchMode {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:84]: Let's write doc-comments (`///`) for this function and explain the motivation behind this comparison logic.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 11]
```diff
@@ -75,9 +114,17 @@ pub struct BoundedWindowAggExec {
...
+    pub partition_search_mode: PartitionSearchMode,
+    /// Partition by indices that define ordering
+    // For example, if input ordering is ORDER BY a, b and window expression
+    // contains PARTITION BY b, a; `ordered_partition_by_indices` would be 1, 0.
+    // Similarly, if window expression contains PARTITION BY a, b; then
+    // `ordered_partition_by_indices` would be 0, 1.
+    // See `get_ordered_partition_by_indices` for more details.
     ordered_partition_by_indices: Vec<usize>,
+    /// Whether the source is unbounded
+    pub source_unbounded: bool,
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:127]: I don't like having this attribute here. It is never used within the operator code, you only seem to be setting this in `PipelineFixer` and then later on using it in `EnforceSorting`. Let’s explore how we can track the boundedness information during the bottom-up traversal in `EnforceSorting` (maybe in `PlanWithCorrespondingSort`) and remove this attribute from this struct.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 12]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
     }
 }
 
+/// Trait that specifies how we search for (or calculate) partitions. It has two
+/// implementations: [`SortedSearch`] and [`LinearSearch`].
+trait PartitionSearcher: Send {
+    /// This method constructs output columns using the result of each window expression.
+    fn calculate_out_columns(
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:377]: We should have a longer doc-comment here and give more context by talking about the arguments of this function and its return value.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 13]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+        )?;
+
+        let n_window_col = window_agg_states.len();
+        let mut new_columns = vec![vec![]; n_window_col];
+        // Size of all_indices can be at most input_buffer.num_rows():
+        let mut all_indices = UInt32Builder::with_capacity(input_buffer.num_rows());
+        for (row, indices) in partition_output_indices {
+            let length = indices.len();
+            for (idx, window_agg_state) in window_agg_states.iter().enumerate() {
+                let partition = &window_agg_state[&row];
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:516]: It seemed to me that `row` would always exist in `window_agg_state`, so I removed the pesky accessor/err closure and used direct access. Please verify if this indeed true or not.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 14]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+                let values = partition.state.out_col.slice(0, length).clone();
+                new_columns[idx].push(values);
+            }
+            let partition_batch_state = &mut partition_buffers[&row];
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:520]: It seemed to me that `row` would always exist in `partition_buffers`, so I removed the pesky accessor/err closure and used direct access. Please verify if this indeed true or not.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 15]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+                if state.n_row_result_missing > 0 {
+                    break;
+                }
+            }
+            counts.push(per_partition_out_results);
+            cur_window_expr_out_result_len
+        });
+        argmin(out_col_counts).map_or(0, |(min_idx, minima)| {
+            for (row, count) in counts.swap_remove(min_idx).into_iter() {
+                let partition_batch = &mut partition_buffers[row];
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:829]: It seemed to me that `row` would always exist in `partition_buffers`, so I removed the pesky accessor/err closure and used direct access. Please verify if this indeed true or not.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 16]
```diff
@@ -572,28 +1033,25 @@ impl SortedPartitionByBoundedWindowStream {
...
         // Retract no longer needed parts during window calculations from partition batch:
         for (partition_row, n_prune) in n_prune_each_partition.iter() {
-            let partition_batch_state = self
-                .partition_buffers
-                .get_mut(partition_row)
-                .ok_or_else(err)?;
-            let batch = &partition_batch_state.record_batch;
-            partition_batch_state.record_batch =
-                batch.slice(*n_prune, batch.num_rows() - n_prune);
+            let pb_state = &mut self.partition_buffers[partition_row];
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:1038]: It seemed to me that `partition_row` would always exist in `partition_buffers`, so I removed the pesky accessor/err closure and used direct access. Please verify if this indeed true or not.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 17]
```diff
@@ -607,51 +1065,46 @@ impl SortedPartitionByBoundedWindowStream {
...
                 },
             ) in partition_window_agg_states
             {
-                if running_length < n_out {
-                    let n_to_del = min(out_col.len(), n_out - running_length);
-                    let n_to_keep = out_col.len() - n_to_del;
-                    *out_col = out_col.slice(n_to_del, n_to_keep);
-                    running_length += n_to_del;
-                }
+                let partition_batch = &mut self.partition_buffers[partition_key];
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:1086]: It seemed to me that `partition_key` would always exist in `partition_buffers`, so I removed the pesky accessor/err closure and used direct access. Please verify if this indeed true or not.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 18]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+
+impl LinearSearch {
+    /// Initialize a new [`LinearSearch`] partition searcher.
+    fn new(ordered_partition_by_indices: Vec<usize>) -> Self {
+        LinearSearch {
+            input_buffer_hashes: VecDeque::new(),
+            random_state: Default::default(),
+            ordered_partition_by_indices,
+            row_map_batch: RawTable::with_capacity(0),
+            row_map_out: RawTable::with_capacity(0),
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:607]: Does giving these some small (e.g. 256) initial capacity help with performance at all? Probably not, but let's still do a quick experiment to be sure.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 19]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+        }
+    }
+
+    fn evaluate_partition_by_column_values(
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:611]: Let's write explanatory doc-comments (`///`) for this function.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 20]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+            .map(|item| match item.evaluate(record_batch)? {
+                ColumnarValue::Array(array) => Ok(array),
+                ColumnarValue::Scalar(scalar) => Err(DataFusionError::Plan(format!(
+                    "Sort operation is not applicable to scalar value {scalar}"
+                ))),
+            })
+            .collect()
+    }
+
+    fn get_per_partition_indices(
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:628]: Let's write explanatory doc-comments (`///`) for this function.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 21]
```diff
@@ -272,6 +370,470 @@ impl ExecutionPlan for BoundedWindowAggExec {
...
+        // res stores PartitionKey and row indices (indices where these partition occurs in the `batch`) for each partition.
+        let mut result: Vec<(PartitionKey, Vec<u32>)> = vec![];
+        for (hash, row_idx) in batch_hashes.into_iter().zip(0u32..) {
+            let entry = self.row_map_batch.get_mut(hash, |(_, group_idx)| {
+                // We can safely get the first index of the partition indices
+                // since partition indices has one element during initialization.
+                let row = get_row_at_idx(columns, row_idx as usize).unwrap();
+                // Handle hash collusions with an equality check:
+                row.eq(&result[*group_idx].0)
+                // equal_rows(row_idx, res[*group_idx].1[0], columns, columns, true).unwrap()
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:647]: Why is this here? Did you plan to move the `equal_rows` function from the join code into utils and use it here too? Did you abandon that idea later on? Why? Let's discuss.

**Review by ozankabak — APPROVED** [review]
I sent a refactor that improves the code structure, removes a lot of clones and makes the code a little more functional in terms of style.

I left a bunch of questions and TODOs in my comments (16 of them!). GitHub hides some of them in this view so please be careful and go through all. We will finalize/send upstream after our meeting tomorrow.

File: datafusion/core/src/physical_optimizer/pipeline_fixer.rs (hunk 6 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/pipeline_fixer.rs:212]: Yes it was the intend. Alternatively, we could do `is_unbounded` check in `EnforceSorting` (recursive call).

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs (hunk 8 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/sort_enforcement.rs:658]: It should be "... is ...". I have re-worded the comment.

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs (hunk 10 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:84]: No need for it

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs (hunk 12 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:377]: Updated doc comment

**Comment by ozankabak** [comment]
Merged upstream, closing.
